File: Examples/sample.cs
// Prism C# sample.
//
// Exercises:
//   - file-scoped namespaces (C# 10+)
//   - records, primary constructors, init-only setters
//   - pattern matching with switch expressions
//   - LINQ + lambdas
//   - async/await, Task<T>
//   - attributes ([Obsolete], custom), preprocessor (#region, #if)
//   - verbatim strings (@"..."), interpolated ($"..."), interpolated verbatim ($@"...")
//   - nullable reference types

#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Prism.Examples;

/// <summary>Lightweight greeter app demonstrating modern C# features.</summary>
public sealed class GreeterApp
{
    #region Constants

    public const double Pi         = 3.14159_26535;
    public const long   HexMask    = 0xDEAD_BEEFL;
    public const int    Octal      = 0b0_1111_0000_0000;
    public const int    Binary     = 0b1010_1010;
    public const float  Scientific = 6.022e23f;
    public const char   Newline    = '\n';

    #endregion

    /* ------------------------------------------------------------------
     * Records + sealed inheritance
     * ------------------------------------------------------------------ */

    public abstract record Shape
    {
        public abstract double Area { get; }
    }

    public sealed record Circle(double Radius) : Shape
    {
        public override double Area => Pi * Radius * Radius;
    }

    public sealed record Rect(double Width, double Height) : Shape
    {
        public override double Area => Width * Height;
    }

    /* ------------------------------------------------------------------
     * Custom attribute
     * ------------------------------------------------------------------ */

    [AttributeUsage(AttributeTargets.Method)]
    private sealed class ExperimentalAttribute : Attribute
    {
        public string Reason { get; }
        public ExperimentalAttribute(string reason) => Reason = reason;
    }

    [Experimental("API may change")]
    public static T? LargestBy<T>(IEnumerable<T> items, Func<T, double> selector)
        where T : class
    {
        return items.OrderByDescending(selector).FirstOrDefault();
    }

    /* ------------------------------------------------------------------
     * Pattern matching
     * ------------------------------------------------------------------ */

    public static string Describe(Shape s) => s switch
    {
        Circle { Radius: > 0 } c => $"circle r={c.Radius:F2}",
        Rect (var w, var h)      => $"rect {w}x{h}",
        _                        => "unknown shape",
    };

    /* ------------------------------------------------------------------
     * Async showcase
     * ------------------------------------------------------------------ */

    [Obsolete("use FetchSquaresAsync instead")]
    public static async Task<int> FetchSquareAsync(int value)
    {
        await Task.Yield();
        return value * value;
    }

    public static async Task Main(string[] args)
    {
        var banner = $@"
Prism C# sample
===============
version : 1.0
mask    : 0x{HexMask:X8}
arg     : {(args.Length > 0 ? args[0] : "<none>")}
";
        Console.Write(banner);

        Shape[] shapes =
        {
            new Circle(2.0),
            new Rect(3.0, 4.0),
        };

        foreach (var s in shapes)
        {
            Console.WriteLine(Describe(s));
        }

        var largest = LargestBy(shapes, s => s.Area);
        Console.WriteLine($"largest area: {largest?.Area ?? double.NaN}");

        // LINQ + interpolation
        var sumOfSquares = Enumerable.Range(1, 5).Select(x => x * x).Sum();
        Console.WriteLine($"sum of squares (1..5) = {sumOfSquares}");

        // verbatim string
        var path = @"C:\Users\Default\AppData\Local\Prism\config.json";
        Console.WriteLine($"verbatim path: {path}");

#if DEBUG
        Console.Error.WriteLine("(debug build)");
#endif
        await Task.CompletedTask;
    }
}

File: Examples/sample.java
// Prism Java sample.
//
// Exercises:
//   - package + imports
//   - annotations (@Override, @Deprecated, custom)
//   - records, sealed types, switch expressions
//   - generics, bounded type parameters
//   - text blocks ("""...""")
//   - lambdas + streams

package com.prism.examples;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

public final class GreeterApp {

    /* ---------------------------------------------------------------------
     * Constants — every numeric form Java supports.
     * --------------------------------------------------------------------- */

    public static final double  PI         = 3.14159_26535;
    public static final long    HEX_MASK   = 0xDEAD_BEEFL;
    public static final int     OCTAL      = 0755;
    public static final int     BINARY     = 0b1010_1010;
    public static final float   SCIENTIFIC = 6.022e23f;
    public static final char    NEWLINE    = '\n';
    public static final char    UNICODE    = 'é';

    /* ---------------------------------------------------------------------
     * Sealed type hierarchy + records
     * --------------------------------------------------------------------- */

    public sealed interface Shape permits Circle, Rectangle, Triangle {
        double area();
    }

    public record Circle(double radius) implements Shape {
        @Override public double area() { return PI * radius * radius; }
    }

    public record Rectangle(double width, double height) implements Shape {
        @Override public double area() { return width * height; }
    }

    public record Triangle(double base, double height) implements Shape {
        @Override public double area() { return 0.5 * base * height; }
    }

    /* ---------------------------------------------------------------------
     * Custom annotation
     * --------------------------------------------------------------------- */

    @Deprecated(since = "1.0", forRemoval = false)
    public @interface Experimental {
        String reason() default "";
    }

    @Experimental(reason = "API not finalized")
    public static <T extends Shape> Optional<T> largest(List<T> shapes) {
        return shapes.stream().max((a, b) -> Double.compare(a.area(), b.area()));
    }

    /* ---------------------------------------------------------------------
     * Greeter using switch expression on the sealed hierarchy
     * --------------------------------------------------------------------- */

    public static String describe(Shape s) {
        return switch (s) {
            case Circle c           -> "circle (r=%.2f)".formatted(c.radius());
            case Rectangle r        -> "rect %dx%d".formatted((int) r.width(), (int) r.height());
            case Triangle t         -> "triangle (b=" + t.base() + ", h=" + t.height() + ")";
        };
    }

    /* ---------------------------------------------------------------------
     * main
     * --------------------------------------------------------------------- */

    public static void main(String[] args) throws Exception {
        var banner = """
                Prism Java sample
                =================
                version : 1.0
                mask    : 0x%08X
                """.formatted(HEX_MASK);
        System.out.print(banner);

        List<Shape> shapes = List.of(
            new Circle(2.0),
            new Rectangle(3.0, 4.0),
            new Triangle(5.0, 6.0)
        );

        var summary = shapes.stream()
            .map(GreeterApp::describe)
            .collect(Collectors.joining(", "));
        System.out.println(summary);

        largest(shapes).ifPresent(s -> System.out.println("largest area: " + s.area()));

        CompletableFuture
            .supplyAsync(() -> "hello from a future")
            .thenAccept(System.out::println)
            .join();
    }
}
